### DISCUSSION TITLE
How can I use a specific Mac / iOS API such as `EventKit`?

### DISCUSSION THREAD

**Discussion by colmugx** [AvaloniaUI/Avalonia#14017] [Q&A] [answered]
I am a Avalonia and dotnet beginner, and I would like to know how should i invoke platform-specific APIs, such as `EventKit` on iOS/MacOS.

I have read alls with search, but cannot find any examples of using platform-specific APIs. This is causing me a lot of confusion.

(btw, Flutter's documentation about specific platforms code is clear :(, i'm sorry about stinking thinking )

**Comment by thevortexcloud** [comment]
Assuming Microsoft has implemented them, use the Mac/iOS TFMs.

https://learn.microsoft.com/en-us/dotnet/standard/frameworks

The platform specific TFMs are supersets of the generic TFMs that add lots of platform specific APIs. Avalonia itself does not really make much of a distinction between platforms beyond what happens during setup.

If Microsoft however has not implemented them, you will likely need to do it yourself/find someone else who has already done it. But it sounds like they have.

https://learn.microsoft.com/en-us/dotnet/api/eventkit?view=xamarin-ios-sdk-12

To also help clarify things: Avalonia is purely a UI framework. It's not a general app framework. The .NET runtime typically provides that functionality.

**Reply by colmugx** [reply]
correctly! There are projects likes `Xamarin.Mac` that have done this work.

However, it seems like this introduces something called `Xamarin`, and I'm not sure if in the dotnet world, the encapsulation for the Mac platform is synonymous with `Xamarin.Mac`. Maybe I can reference this project in my Avalonia project if it was?

**Reply by thevortexcloud** [reply]
So, Xamarin.Mac has technically be deprecated. Which is fairly misleading since all Microsoft did was rename it and merge it into the Mac specific TFM (their own documentation which I linked shows that). As such you no longer directly reference `Xamarin.Mac` you instead use the TFM which then pulls it in for you. So just change your target framework in your project file to `net8.0-macos`. If you use the cross platform template Avalonia ships with it will setup a solution that should give you a general idea how to manage it all. That said it wont make a Mac Specific project. But it will make a iOS, android, web assembly, and desktop one. You can technically multi target in a single project, but it's fairly annoying to manage.

To also clarify again: Xamarin is a project Microsoft bought that made .NET run on many, many devices. Microsoft has just renamed most of the core Xamarin projects now but they still support them in some form. EG MonoDevelop became Visual Studio for Mac. Although they are about to kill that off.

**Reply by colmugx** [reply]
Excited. Thank you for replying.

At the same time, I found comments in the Desktop.fsproj created by default in Avalonia.

```
<!--If you are willing to use Windows/MacOS native APIs you will need to create 3 projects.
        One for Windows with net8.0-windows TFM, one for MacOS with net8.0-macos and one with net8.0 TFM for Linux.-->
```

But i modified `TargetFramework` to `net8.0-macos`, and got error: `0.0 is not a valid TargetPlatformVersion for macos. Valid versions include: None`

**Reply by thevortexcloud** [reply]
> At the same time, I found comments in the Desktop.fsproj created by default in Avalonia.

As I mentioned, what this comment says is not technically true. You _can_ have everything in a single project by multi targeting different TFMs. It's just fairly annoying to manage. But there are valid reasons why someone would want to do it.

In any case: Some TFMs let you specify a specific OS version you are targeting. Can you show me your whole project file so I can check it?

**Reply by colmugx** [reply]
> In any case: Some TFMs let you specify a specific OS version you are targeting. Can you show me your whole project file so I can check it?

Sorry I can't. I'm in the exploration phase, and not any changes. 

However, I have made some attempts (following the default created of MAUI), `net8.0-maccatalyst` can be used as a TargetFramework but Avalonia doesn't seem to handling, it exception.

**Reply by thevortexcloud** [reply, marked as answer]
`maccatalyst` and `macos` are different TFMs. What the actual functional difference is though I am not sure (I don't program for Macs). 

> Sorry I can't. I'm in the exploration phase, and not any changes.

Fair enough. Have you checked if you have the MacOS workload installed? You can do that with `dotnet workload install macos` in a terminal.

**Reply by colmugx** [reply]
> Fair enough. Have you checked if you have the MacOS workload installed? You can do that with dotnet workload install macos in a terminal.

Oops...  I didn't know about that. 

and then I was installed it and changed `TargetFramework` to `net8.0-macos`. It seems ok but.

```
Error  : WinExe is not a valid output type for macOS
```

Here is project file's snippets:

```xml
<PropertyGroup>
      <OutputType>WinExe</OutputType>
      <!--If you are willing to use Windows/MacOS native APIs you will need to create 3 projects.
      One for Windows with net8.0-windows TFM, one for MacOS with net8.0-macos and one with net8.0 TFM for Linux.-->
      <TargetFramework>net8.0-macos</TargetFramework>
      <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
</PropertyGroup>
```

Feels like I've made a little progress. Thank you so much for replying.

**Reply by thevortexcloud** [reply]
I think the output type needs to be `Win` but I am not sure about that.

**Reply by colmugx** [reply]
Thank you so much for help. I think I have an obligation to share something to you.

Ok First, I will tell you what I have been doing recently.

1. I converted the `F#` to `C#` it also for the sake of easier communication.

2. I installed the `macos workload` and changed the `OutputType` to `Exe`, it's ok.

3. I added some configurations, referencing the projects created by default with`macos workload`.

It can run well at least. Here is my project file's snippet:
```xml
<TargetFramework>net8.0-macos</TargetFramework>
<OutputType>Exe</OutputType>
<Nullable>enable</Nullable>
<ImplicitUsings>true</ImplicitUsings>
<BuiltInComInteropSupport>true</BuiltInComInteropSupport>
<SupportedOSPlatformVersion>10.15</SupportedOSPlatformVersion>
```

Now I have some new problems:

currently, I have two projects, `HelloWorld` is with `net8.0` and `Avalonia`, `HelloWorld.Desktop` with `net8.0-macos`. That's not problem.

But I don't know how to call the platform-specific code I've already written in `HelloWorld.Desktop` from `HelloWorld`.

Or I was created new Platform Binding Project, I don't know how to refer. Maybe it's a silly question. Should I get a DI or something?

**Reply by thevortexcloud** [reply]
There are several ways to do this. The easiest (and also easiest to scale) though is to just use DI. I wrote up a fairly minimal C# example awhile back for someone who also was trying to run Avalonia cross platform (including on a Mac) and more or less do the same thing [here](https://github.com/AvaloniaUI/Avalonia/discussions/13455#discussioncomment-7464329).

Typically though you throw as much as you possible can into your shared project (including your service definitions/interfaces). If you don't need to support multiple platforms, you can just use a single project/use a single TFM for all your projects.

**Reply by colmugx** [reply]
Thank you so much.

Well. I will tell you what I have been doing recently.
1. I used Microsoft.Extensions.DependencyInjection.
2. In the `HelloWorld`, I created a `Service` and `Interface`. I created a `CollectionService` as a `Singleton` and `IEventKit`. Then in `HelloWorld.Desktop`, I used `CollectionService` and implemented `IEventKit`, followed by AddSingleton. This may look a bit messy, but it seems to work.

In macOS development, we often use an entitlements file to specify certain entitlements, such as `App Sandbox` and `Calendars`. It seems like the entitlements file in .NET is called `Entitlements.plist`? At least, I see it in the default created macOS project. But it seems like this file may not be taking effect, or I'm not sure what is happening. 

I can confirm that if I follow the same steps in Swift development, this works fine. And I also tried directly creating a new solution, and creating a macOS application in .NET/.NET Core. Following the same steps for development, it really doesn't work.

Anyway, you have helped me a lot. I can't express how grateful I am for your time. Thank you.

**Reply by thevortexcloud** [reply]
> and implemented IEventKit

It almost sounds like you are trying to target Macs only? In which case I would personally not even bother having two projects. That interface name implies you are just indirectly coupling everything to `EventKit` rather than something that `EventKit` provides functionality for. Interfaces generally should be fairly generic. The implementation can be much more specific.

> At least, I see it in the default created macOS project. But it seems like this file may not be taking effect, or I'm not sure what is happening.

I know for a fact Avalonia [does override some plist values at runtime.](https://github.com/AvaloniaUI/Avalonia/discussions/13568#discussioncomment-7587629) Although which ones it does I can't say. However if it's not working at all then you possibly need to tell the runtime to use the file? Google shows me [this](https://learn.microsoft.com/en-us/dotnet/maui/ios/entitlements?view=net-maui-8.0) MAUI article for configuring an iOS project to use it. It apparently just adds this XML to your project file

```xml
 <CodesignEntitlements>Platforms\iOS\Entitlements.plist</CodesignEntitlements>
```
In this thread it looks like it's being used manually via the CLI.

https://github.com/AvaloniaUI/Avalonia/issues/9764

However the thread is a year old so perhaps some things have changed since then.

> Anyway, you have helped me a lot. I can't express how grateful I am for your time. Thank you.

You are welcome! Glad I could help. I wish I could give better advice but I really have no experience developing for Macs.

**Reply by colmugx** [reply]
> It almost sounds like you are trying to target Macs only?

No. Actually the interface should be `IReminder`, and I will implement this interface separately on Windows and Mac.

Of course, since our primary platform is Mac, and we hope to synchronize Reminders through `EventKit` yea.

Anyway, thank you for your a lot of time!
